docs: add instructions and details to contributors guide

lance opened the pull request:
* Added detailed instructions for outside contributors.
* Added tips for maintainers
* Documented branching strategy

Fixes: https://github.com/cloudevents/sdk-javascript/issues/85
Fixes: https://github.com/cloudevents/sdk-javascript/issues/72

Signed-off-by: Lance Ball <lball@redhat.com>

lance commented:
By way of example, I have submitted a [pull request](https://github.com/cloudevents/sdk-javascript/pull/106) for backporting on the `v1.x.y` branch. This pull request is a cherry pick of the [commit](https://github.com/cloudevents/sdk-javascript/commit/3a063d72451d1156df8fe9c3499ef1e81e905060) where the fix landed on master. In the commit message I've included a reference to this and the PR that the fix was proposed in. This allows us to cut a `v1.1.0` bug fix release from the `v1.x.y` branch without having to worry about potential breaking changes on master.

I've also, inadvertently, provided examples of force pushing minor changes to my fork. lol

Currently our CI is configured to build all branches, but ultimately I think it should only be building `master`, these version branches, and pull requests.

duglin commented on CONTRIBUTING.md:
````diff
@@ -2,61 +2,159 @@
-- [Gitflow](https://nvie.com/posts/a-successful-git-branching-model/) by @nvie
+```console
+git clone https://github.com/mygithuborg/sdk-javascript.git
+git remote add upstream https://github.com/cloudevents/sdk-javascript.git
````
need a 'cd sdk-javascript' before this line

helio-frota replied:
good catch

helio-frota commented on CONTRIBUTING.md:
````diff
@@ -2,61 +2,159 @@
+```console
+git fetch upstream
+git reset --hard upstream/master
+git co -b 48-fix-http-agent-error
````
Minor change to use `checkout` instead of `co`.

I don't have the co alias so it gets confusing:
```
09:07 $ git co
git: 'co' is not a git command. See 'git --help'.
```

lance commented on CONTRIBUTING.md:
```diff
@@ -2,61 +2,159 @@
+* Stay on top of your pull requests. PRs that languish for too long can become difficult to merge.
+* Work from your own fork. As you are making contributions to the project, you should be working from your own fork just as outside contributors do. This keeps the branches in github to a minimum and reduces unnecessary CI runs.
+* Try to proactively label issues with backport labels if it's obvious that a change should be backported to previous releases.
+* When landing pull requests, if there is more than one commit, try to squash into a single commit.
```
Well, we have both rebase and squash merging enabled on the repository. I think as a rule we should be using squash, but there may be times when multiple commits in a PR make sense. So I don't think we should change the settings to _only_ allow squash merges. That's why I said "try to squash". But maybe I should be more explicit.

lance commented:
So I raised this in the #cloudeventssdk Slack channel and got zero engagement. The fact that the Go SDK (or any of the others) don't have contributor's guides doesn't mean in my mind that it's OK not to. It's pretty standard practice, and if I were contributing to those repos, I would make the same suggestion.

Do you have some suggested contribution guides that you think we should link to?

Are there specific concerns you have with my proposed workflow suggestions, other than "people don't want to read that"? Is it just the `git` workflow that you have problems with? Or is it the whole thing?

I loosely based the workflow portion of this on the [Node.js contributor's guide](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#setting-up-your-local-environment), which is actually much more detailed and longer than what I have written. Of course I realize that it's a much bigger project as well, so our processes don't need to be as strict. But don't you think it would be nice to have a standard set of recommendations for contributors to follow? It's not like I'm saying nothing will land unless you follow these steps. It just makes it easy to point to an existing doc when people have questions about how to go about it. And it makes the commit log appear clean and consistent. 

I want to keep the barriers to contribution low. I want first time contributors to feel confident in their work, so that it's less scary to submit a PR to a new org. I think providing _more_ - not less - information, tips, guidance is a way to achieve that.

lance commented on CONTRIBUTING.md:
```diff
@@ -2,61 +2,160 @@
+Code style for this module is maintained using [`eslint`](https://www.npmjs.com/package/eslint).
+When you run tests with `npm test` linting is performed first. If you want to
+check your code style for linting errors without running tests, you can just
+run `npm run lint`. If there are errors, you can usually fix them automatically
```
Yes - linting is run as a pre-task whenever `npm test` is run. https://github.com/cloudevents/sdk-javascript/blob/master/package.json#L9

lance commented on CONTRIBUTING.md:
```diff
@@ -2,61 +2,160 @@
+## Maintainer's Guide
 
-### JavaScript Style Guide
+Here are a few tips for repository maintainers.
```
Good suggestions. I will add these or something similar. Yesterday I added a `stalled` label for issues and PRs which I think can be used in this case.

lance commented:
@slinkydeveloper suggestions applied. Thanks for taking a look.

@grant I removed the branch management stuff - well, pared it way down - and moved it to the "Contributors" section. Also, not sure if you saw my comment earlier:

> Do you have some suggested contribution guides that you think we should link to?

> Are there specific concerns you have with my proposed workflow suggestions, other than "people don't want to read that"? Is it just the git workflow that you have problems with? Or is it the whole thing? Could you be specific about the sections you object to? I'm happy to accept feedback, but I'm not sure what you mean by simplifying.

grant commented:
@lance Thanks for the improvements. I'm hoping for a simple guide of vital need-to-know information.

Here are some simple guides that we use at Google:

- https://github.com/google/new-project/blob/master/docs/contributing.md
- https://github.com/googleapis/google-api-nodejs-client/blob/master/CONTRIBUTING.md

I think a common standard among all SDKs would be ideal as we care about SDK all major languages. A paragraph about a language-specific contribution sounds fine though.

lance commented:
@grant OK, I can make those changes - but would like to pull the git instructions out and include a linked step by step guide for those just getting started. I know that when I began contributing to other OSS projects in the past, I personally found these kinds of guides helpful. Sound like a reasonable compromise?

grant commented:
@lance Sounds fine.
- Things like copying the exact text of developercertificate.org instead of linking it doesn't provide much use.
- Nifty guides that aren't found elsewhere around git would be helpful.

slinkydeveloper commented:
Great job @lance, can you port the same guide in other sdks?

lance commented:
@slinkydeveloper sure can - will put it on the todo list.

lance commented:
OK - the PR guidelines were pulled into a separate document, the maintainer tips were pulled out into a second document, and the main contributing doc was greatly simplified. Can I get an approval on this @grant?

grant commented:
> OK - the PR guidelines were pulled into a separate document, the maintainer tips were pulled out into a second document, and the main contributing doc was greatly simplified. Can I get an approval on this @grant?

Thanks for the improvements Lance.
Right now I'm looking at all the CE SDKs and can't really approve this PR unless we want the same guidelines in all 7 CE SDKs. In my opinion, this should be dramatically reduced. For example, we shouldn't include common instructions like how to use git.

I'll remove my review request to unblock you though.

lance commented:
@grant does that mean you are not explicitly blocking this from landing, just registering dissent? I'm not clear.

grant commented:
> @grant does that mean you are not explicitly blocking this from landing, just registering dissent? I'm not clear.

I just don't approve or request changes. Usually it's not good to merge a PR with requested changes. I just dismissed my review.

As stated I don't really agree with adding any more info about `git` than is needed. I'd imagine 3-5 bullet points about contributing. I wouldn't imagine a GitHub user to have to read 7 different contributing guides for each SDK.

So, while I can appreciate `git` guides, style guides, and thank you's in READMEs, it doesn't scale to new developers or hundreds of projects. I'd rather just link a git handbook, use a standard style tool and keep to < 20 lines of code in the README.

lance commented:
Now sdk-go has [adopted these documents](https://github.com/cloudevents/sdk-go/pull/501) - I'm going to land this.
